nemo_skills/inference/prompt/utils.py

```diff
@@ -37,19 +39,92 @@
+        if self.example_dicts is None:
+            self.example_dicts = examples_map.get(self.config.examples_type, [])[: self.config.num_few_shots]
+        if self.context_template is None:
+            self.context_template = context_templates.get(self.config.context_type, "")
+
+    def build_context(self, example_dict: Dict[str, Any]) -> str:
+        """Builds the context string based on the example dictionary."""
+        context = self.context_template.format(**example_dict)
+        return context
+
+    def build_filled_example(self, example_dict: Dict[str, Any]) -> str:
+        """Builds a filled example string based on the example dictionary."""
+        context = self.build_context(example_dict)
+        return self.config.few_shot_examples.template.format(context=context, **example_dict)
+
+    def build_examples(self) -> str:
+        """Builds all examples string concatenated by delimiter."""
+        filled_examples = [self.build_filled_example(example) for example in self.example_dicts]
+        examples = (
+            self.config.few_shot_examples.delimiter.join(filled_examples) + self.config.few_shot_examples.delimiter
+            if filled_examples
+            else ""
+        )
+        context = self.build_context(self.input_dict)
+        user = self.config.user.format(examples=examples, context=context, **self.input_dict)
+        return user
+
+    def build_structured_examples(self) -> List[Dict[str, str]]:
+        """Builds a structured representation of the prompt."""
+        structured_prompt = [{"role": "system", "content": self.config.system}] if self.config.system else []
+        structured_prompt.append({"role": "user", "content": self.build_examples()})
+        if self.generated_solution:
+            structured_prompt.append({"role": "assistant", "content": self.generated_solution})
+        return structured_prompt
+
+    def __add__(self, other: str):
+        return self.generated_solution + other
+
+    def __iadd__(self, other: str):
+        self.generated_solution += other
+        return self
+
+    def __str__(self) -> str:
+        """Returns the complete prompt string representation."""
+        prompt = self.config.prompt_template.format(
+            system=self.config.system,
+            user=self.build_examples(),
+            generated_solution=self.generated_solution,
+        )
+        return prompt
+
+
 def get_prompt_config(prompt_type: str) -> PromptConfig:
     # reading prompt format from the yaml file, if not running through hydra
-    with open(Path(__file__).parent / f"{prompt_type}.yaml", "rt", encoding="utf-8") as fin:
-        return PromptConfig(**yaml.safe_load(fin))
+    config_path = Path(__file__).parent / f"{prompt_type}.yaml"
+    with open(config_path, "rt", encoding="utf-8") as fin:
+        # Load the YAML file using OmegaConf
+        loaded_config = OmegaConf.load(fin)
+        prompt_config = OmegaConf.structured(PromptConfig(**loaded_config))
```

Comment:
You might need to call OmegaConf.to_object instead or in addition to in order to force all inner dataclasses to be properly built as well